### PR TITLE
Bug fix - cannot construct Index of empty tuples

### LINKED ISSUE #45608: BUG: cannot construct Index of empty tuples
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd
pd.Index([()]) # Fails, the MultiIndex constructor complains that you "must pass non-zero number of levels/codes"
pd.Index([(), None])[:1] # Works
```

### Issue Description

The index constructor has a special case for when the input list consists purely of tuples. However, it's not aware of the special case in MultiIndex that forbids number of levels == 0. Index should avoid trying to construct a MultiIndex in this case.

(Even better, MultiIndex could just support 0-level indexes. Yes, they are trivial in some sense, but they are also perfectly sensible.)

### Expected Behavior

Both lines should evaluate correctly to identical indexes.

### Installed Versions

<details>
INSTALLED VERSIONS
------------------
commit           : 945c9ed766a61c7d2c0a7cbb251b6edebf9cb7d5
python           : 3.8.12.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19042
machine          : AMD64
processor        : Intel64 Family 6 Model 85 Stepping 7, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : English_United Kingdom.1252

pandas           : 1.3.4
numpy            : 1.21.3
pytz             : 2021.3
dateutil         : 2.8.2
pip              : 21.1.1
setuptools       : 58.0.4
Cython           : 0.29.24
pytest           : 6.2.4
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.6.3
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : 3.0.2
IPython          : 7.29.0
pandas_datareader: None
bs4              : 4.10.0
bottleneck       : 1.3.2
fsspec           : 2021.08.1
fastparquet      : None
gcsfs            : None
matplotlib       : 3.4.3
numexpr          : 2.7.3
odfpy            : None
openpyxl         : 3.0.9
pandas_gbq       : None
pyarrow          : 3.0.0
pyxlsb           : None
s3fs             : None
scipy            : 1.6.2
sqlalchemy       : None
tables           : 3.6.1
tabulate         : None
xarray           : 0.19.0
xlrd             : 2.0.1
xlwt             : None
numba            : 0.53.1
</details>

### PR DESCRIPTION
**Pull request by weikhor**
- [x] closes #45608 
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by jreback** [comment]
this is failing a lot of tests.

**Comment by weikhor** [comment]
> this is failing a lot of tests.
Yes. I need to fix codes to include all test cases.

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -542,9 +541,10 @@ def __new__(
                     # 10697
                     from pandas.core.indexes.multi import MultiIndex
 
-                    return MultiIndex.from_tuples(
-                        data, names=name or kwargs.get("names")
-                    )
+                    if all(data):
```
Comment (jreback) [inline pandas/core/indexes/base.py:544]: instead of this you should modify `from_tuples` itself

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/base.py (hunk 1 above)
Comment (weikhor) [reply pandas/core/indexes/base.py]: Will update it.

Comment (weikhor) [reply pandas/core/indexes/base.py]: Have update it.

**Comment by jreback** [comment]
> pandas/core/indexes/multi.py:493: error: List item 0 has incompatible type "List[Any]"; expected "ndarray[Any, Any]"  [list-item]

**Comment by weikhor** [comment]
> > pandas/core/indexes/multi.py:493: error: List item 0 has incompatible type "List[Any]"; expected "ndarray[Any, Any]"  [list-item]

Have resolve

File: pandas/core/indexes/multi.py [hunk 2]
```diff
@@ -487,6 +487,14 @@ def from_arrays(cls, arrays, sortorder=None, names=lib.no_default) -> MultiIndex
                 raise ValueError("all arrays must be same length")
 
         codes, levels = factorize_from_iterables(arrays)
+
+        if all(isinstance(e, tuple) for e in arrays):
```
Comment (mroeschke) [inline pandas/core/indexes/multi.py:491]: Can this just be done in `from_tuples`?

Comment (weikhor) [reply pandas/core/indexes/multi.py:491]: Yes. I think can. Will try it.

File: pandas/core/indexes/multi.py [hunk 3]
```diff
@@ -563,6 +563,20 @@ def from_tuples(
             arrs = zip(*tuples)
             arrays = cast(List[Sequence[Hashable]], arrs)
 
+        if all(isinstance(e, tuple) for e in tuples):
+            if not all(tuples):
```
Comment (mroeschke) [inline pandas/core/indexes/multi.py:567]: Doesn't the above condition satisfy this one?

File: pandas/core/indexes/multi.py [hunk 4]
```diff
@@ -563,6 +563,20 @@ def from_tuples(
...
+                codes = [np.array([0 for _ in range(len(arrays))])]
```
Comment (mroeschke) [inline pandas/core/indexes/multi.py:568]: `np.zeros`

File: pandas/core/indexes/multi.py [hunk 5]
```diff
@@ -563,6 +563,20 @@ def from_tuples(
...
+                _dtype_obj = np.dtype("object")
```
Comment (mroeschke) [inline pandas/core/indexes/multi.py:569]: Can just inline where it's used

File: pandas/core/indexes/multi.py (hunk 4 above)
Comment (weikhor) [reply pandas/core/indexes/multi.py:568]: Done

Comment (weikhor) [reply pandas/core/indexes/multi.py:568]: Done

File: pandas/core/indexes/multi.py (hunk 3 above)
Comment (weikhor) [reply pandas/core/indexes/multi.py]: No. `tuples` can be not list of tuples. For example, based on a test case,  `tuples` can be 
```
tuples = [array([0, 1], dtype=int64), array([0, 0], dtype=int64), array([1, 0], dtype=int64)]
sortorder = None, names = ['2', 'a']
```
from `pandas/tests/strings/test_api.py` , `test_api_per_method` function

File: pandas/core/indexes/multi.py [hunk 6]
```diff
@@ -562,6 +562,18 @@ def from_tuples(
             arrs = zip(*tuples)
             arrays = cast(List[Sequence[Hashable]], arrs)
 
+        if all(isinstance(e, tuple) for e in tuples) and not all(tuples):
+            codes = [np.array([0 for _ in range(len(arrays))])]
```
Comment (mroeschke) [inline pandas/core/indexes/multi.py:566]: `np.zeros`

File: pandas/core/indexes/multi.py [hunk 7]
```diff
@@ -562,6 +562,20 @@ def from_tuples(
             arrs = zip(*tuples)
             arrays = cast(List[Sequence[Hashable]], arrs)
 
+        if len(tuples) != 0:
```
Comment (jreback) [inline pandas/core/indexes/multi.py:565]: this is going to be a considerable perf hit.

can you just look at the arrays?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/multi.py (hunk 7 above)
Comment (weikhor) [reply pandas/core/indexes/multi.py]: Hi I not able to use arrays. If `tuples =[ ( ), ( ) ]`. 
```
elif isinstance(tuples, list):
     arrays = list(lib.to_object_array_tuples(tuples).T)
```
The arrays will equal to `[ ]` after this section

File: pandas/core/indexes/multi.py [hunk 8]
```diff
@@ -545,7 +545,16 @@ def from_tuples(
         elif is_iterator(tuples):
             tuples = list(tuples)
         tuples = cast(Collection[Tuple[Hashable, ...]], tuples)
-
+        if len(tuples) and all((isinstance(e, tuple) and not e) for e in tuples):
```
Comment (mroeschke) [inline pandas/core/indexes/multi.py:548]: Nit: `(isinstance(e, tuple) and not e)` doesn't need the outer parenthesis

**Review by mroeschke — COMMENTED** [review]
Looks pretty good. Just needs a whatsnew note

File: pandas/core/indexes/multi.py (hunk 8 above)
Comment (jreback) [inline pandas/core/indexes/multi.py:548]: add a blank line then a comment that this is handling the empty tuple case

Comment (weikhor) [reply pandas/core/indexes/multi.py]: Done

File: doc/source/whatsnew/v1.5.0.rst [hunk 9]
```diff
@@ -485,7 +485,7 @@ MultiIndex
 - Bug in :meth:`DataFrame.loc` raising when slicing a :class:`MultiIndex` with a negative step size and slicing a non-int labeled index level (:issue:`46156`)
 - Bug in :meth:`Series.to_numpy` where multiindexed Series could not be converted to numpy arrays when an ``na_value`` was supplied (:issue:`45774`)
 - Bug in :class:`MultiIndex.equals` not commutative when only one side has extension array dtype (:issue:`46026`)
--
+- bug in :meth:`MultiIndex.from_tuples` cannot construct Index of empty tuples (:issue:`45608`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v1.5.0.rst:488]: Nit: `bug` -> `Bug`

**Review by mroeschke — COMMENTED** [review]
Small comment otherwise LGTM

**Comment by weikhor** [comment]
Thank for review.

**Review by mroeschke — APPROVED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @weikhor

**Pull request merged**